lib/ansible/module_utils/oracle/oci_utils.py

```diff
@@ -0,0 +1,1982 @@
+    :param kwargs_delete: Dictionary of arguments for delete function delete_fn. e.g. {"vcn_id": module.params["id"]}
+    :param module: Instance of AnsibleModule.
+    :param states: List of lifecycle states to watch for while waiting after delete_fn is called. If nothing is passed,
+                   defaults to ["TERMINATED", "DETACHED", "DELETED"].
+    :param process_work_request: Whether a work request is generated on an API call and if it needs to be handled.
+    :return: A dictionary containing the resource & the "changed" status. e.g. {"vcn":{x:y}, "changed":True}
+    """
+
+    states_set = set(["DETACHING", "DETACHED", "DELETING", "DELETED", "TERMINATING", "TERMINATED"])
+    result = dict(changed=False)
+    result[resource_type] = dict()
+    try:
+        resource = to_dict(call_with_backoff(get_fn, **kwargs_get).data)
+        if resource:
+            if "lifecycle_state" not in resource or resource["lifecycle_state"] not in states_set:
+                response = call_with_backoff(delete_fn, **kwargs_delete)
+                if process_work_request:
+                    wr_id = response.headers.get("opc-work-request-id")
+                    get_wr_response = call_with_backoff(
+                        client.get_work_request, work_request_id=wr_id
+                    )
+                    result["work_request"] = to_dict(
+                        wait_on_work_request(client, get_wr_response, module)
+                    )
+                    # Set changed to True as work request has been created to delete the resource.
+                    result["changed"] = True
+                    resource = to_dict(call_with_backoff(get_fn, **kwargs_get).data)
+                else:
+                    _debug("Deleted {0}, {1}".format(resource_type, resource))
+                    result["changed"] = True
+
+                    if wait_applicable and module.params.get("wait", None):
+                        if states is None:
+                            states = (
+                                module.params.get("wait_until")
+                                or DEFAULT_TERMINATED_STATES
+                            )
+                        try:
+                            wait_response = oci.wait_until(
+                                client,
+                                get_fn(**kwargs_get),
+                                evaluate_response=lambda r: r.data.lifecycle_state
+                                in states,
+                                max_wait_seconds=module.params.get(
+                                    "wait_timeout", MAX_WAIT_TIMEOUT_IN_SECONDS
+                                ),
+                                succeed_on_not_found=True,
+                            )
+                        except MaximumWaitTimeExceeded as ex:
+                            module.fail_json(msg=str(ex))
+                        except ServiceError as ex:
+                            if ex.status != 404:
+     
```

Comment:
What do you need this function for? Ansible already handles aliases, there's no need to do this manually.

Reply:
@felixfontein can you please clarify your comment? I don't see any aliases at the comment line.

Reply:
(GitHub is showing this with the wrong code snippet in the discussion overview. This comment is about the `_get_user_provided_value` function.)

Reply:
@felixfontein  The purpose for _get_user_provided_value function is to allow us to compare the attributes returned from the cloud with the attributes provided by the user. If the values are different, we send an update request to the cloud. If we remove this function we have to change the naming convention that we use in all modules.

Currently we use this naming convention 
name=dict(type=“str”, required=False, aliases=[“display_name”]), 
Without this function, we need to flip the naming convention to to match the SDK attributes as the main values:
display_name=dict(type=“str”, required=False, aliases=[“name”])

We prefer to keep the current convention because it makes the modules' attributes more user friendly and more consistent. As you see, we are already using Ansible aliasing and this function is not used to handle aliasing or to re-implement an exciting functionality.

Reply:
Sounds ok.